### PR TITLE
Fixed Issue Preventing Agg on RollingGroupBy Objects

### LINKED ISSUE #15072: BUG/ENH: groupby.rolling.agg / Column(s) low already selected
#### GroupBy.rolling.agg failed with Column(s) low already selected

```python
data = pd.DataFrame({"stock": [1, 1, 1, 2, 2, 2, 2], "low": [10, 20, 30, 10, 30, 40, 80]})
data.set_index("stock", inplace=True)
# failed
data.groupby(level="stock").rolling(2).agg({"low": {"mean": "mean", "max": "max"}})
# works
data.rolling(2).agg({"low": {"mean": "mean", "max": "max"}})
```

This also hits `groupby().resample.agg`

```python
In [13]: df = pd.DataFrame({"A": pd.to_datetime(['2015', '2017']), "B": [1, 1]})

In [14]: df
Out[14]:
           A  B
0 2015-01-01  1
1 2017-01-01  1

In [15]: df.set_index("A").groupby([0, 0]).resample("AS")
Out[15]: DatetimeIndexResamplerGroupby [freq=<YearBegin: month=1>, axis=0, closed=left, label=left, convention=e, base=0]

In [16]: df.set_index("A").groupby([0, 0]).resample("AS").agg(['sum', 'count'])
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-16-5f1c18a8d4ac> in <module>()
----> 1 df.set_index("A").groupby([0, 0]).resample("AS").agg(['sum', 'count'])

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/resample.py in aggregate(self, arg, *args, **kwargs)
    339
    340         self._set_binner()
--> 341         result, how = self._aggregate(arg, *args, **kwargs)
    342         if result is None:
    343             result = self._groupby_and_aggregate(arg,

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/base.py in _aggregate(self, arg, *args, **kwargs)
    538             return self._aggregate_multiple_funcs(arg,
    539                                                   _level=_level,
--> 540                                                   _axis=_axis), None
    541         else:
    542             result = None

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/base.py in _aggregate_multiple_funcs(self, arg, _level, _axis)
    583                 try:
    584                     colg = self._gotitem(col, ndim=1, subset=obj[col])
--> 585                     results.append(colg.aggregate(arg))
    586                     keys.append(col)
    587                 except (TypeError, DataError):

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/resample.py in aggregate(self, arg, *args, **kwargs)
    339
    340         self._set_binner()
--> 341         result, how = self._aggregate(arg, *args, **kwargs)
    342         if result is None:
    343             result = self._groupby_and_aggregate(arg,

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/base.py in _aggregate(self, arg, *args, **kwargs)
    538             return self._aggregate_multiple_funcs(arg,
    539                                                   _level=_level,
--> 540                                                   _axis=_axis), None
    541         else:
    542             result = None

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/base.py in _aggregate_multiple_funcs(self, arg, _level, _axis)
    582             for col in obj:
    583                 try:
--> 584                     colg = self._gotitem(col, ndim=1, subset=obj[col])
    585                     results.append(colg.aggregate(arg))
    586                     keys.append(col)

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/base.py in _gotitem(self, key, ndim, subset)
    675                        for attr in self._attributes])
    676         self = self.__class__(subset,
--> 677                               groupby=self._groupby[key],
    678                               parent=self,
    679                               **kwargs)

~/Envs/pandas-dev/lib/python3.6/site-packages/pandas/pandas/core/base.py in __getitem__(self, key)
    241         if self._selection is not None:
    242             raise Exception('Column(s) {selection} already selected'
--> 243                             .format(selection=self._selection))
    244
    245         if isinstance(key, (list, tuple, ABCSeries, ABCIndexClass,

Exception: Column(s) B already selected
```

#### Problem description

It's not work as expected

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-53-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.19.2+0.g825876c.dirty
nose: 1.3.7
pip: 9.0.1
setuptools: 27.2.0
Cython: 0.24.1
numpy: 1.11.1
scipy: 0.18.1
statsmodels: 0.6.1
xarray: None
IPython: 5.1.0
sphinx: 1.4.6
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.6.1
blosc: None
bottleneck: 1.1.0
tables: 3.2.3.1
numexpr: 2.6.1
matplotlib: 1.5.3
openpyxl: 2.3.2
xlrd: 1.0.0
xlwt: 1.1.2
xlsxwriter: 0.9.3
lxml: 3.6.4
bs4: 4.5.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.13
pymysql: 0.7.9.None
psycopg2: None
jinja2: 2.8
boto: 2.42.0
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #15072
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

AFAICT the fact that RollingGroupBy could not use `agg` with a list of functions is simply due to the fact that the GroupByMixing it inherits from could not handle the reduction in dimensions that occurs via the normal aggregation functions.

### CODE REVIEW

File: pandas/core/base.py [hunk 1]
```diff
@@ -684,8 +684,14 @@ def _gotitem(self, key, ndim, subset=None):
         # with the same groupby
         kwargs = dict([(attr, getattr(self, attr))
                        for attr in self._attributes])
+
+        if self._groupby.ndim == 1 and self._groupby.obj.name == key:
```
Comment (jreback) [inline pandas/core/base.py]: I hate checks like this :>

does
``groupby = self._groupby[key]`` not just work here?

File: pandas/tests/test_window.py [hunk 2]
```diff
@@ -520,6 +520,39 @@ def test_iter_raises(self, klass):
         with pytest.raises(NotImplementedError):
             iter(obj.rolling(2))
 
+    def test_agg(self):
```
Comment (jreback) [inline pandas/tests/test_window.py]: can be a bit more descriptive here on the test name / expl

File: pandas/tests/test_window.py [hunk 3]
```diff
@@ -520,6 +520,39 @@ def test_iter_raises(self, klass):
...
+            ['B', 80, 90]], columns=['stock', 'low', 'high'])
+
+        index = pd.MultiIndex.from_tuples([
+            ('A', 0), ('A', 1), ('A', 2),
+            ('B', 3), ('B', 4), ('B', 5), ('B', 6)], names=['stock', None])
+        columns = pd.MultiIndex.from_tuples([
+            ('low', 'mean'), ('low', 'max'), ('high', 'mean'),
+            ('high', 'min')])
+        expected = pd.DataFrame([
+            [np.nan, np.nan, np.nan, np.nan],
```
Comment (jreback) [inline pandas/tests/test_window.py]: maye put in TestApi. this should test rolling & expanding (atm we don't have a nice modular structure to do this :<)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/base.py [hunk 4]
```diff
@@ -684,8 +684,14 @@ def _gotitem(self, key, ndim, subset=None):
         # with the same groupby
         kwargs = dict([(attr, getattr(self, attr))
                        for attr in self._attributes])
+
+        try:
+            groupby = self._groupby[key]  # get slice of frame
+        except Exception:
```
Comment (WillAyd) [inline pandas/core/base.py]: Ideally would catch something more descriptive than an `Exception`, but that is all that gets raised:

https://github.com/pandas-dev/pandas/blob/67e6e6fcd19d1d89cb60abc3a78372bc85fd8e29/pandas/core/base.py#L249

Maybe an AttributeError would be more appropriate?

**Comment by WillAyd** [comment]
Test errors have to do with the column sorting of the returned frame. Do we make any guarantees about how that is supposed to be ordered across versions?

File: pandas/core/base.py (hunk 4 above)
Comment (jreback) [reply pandas/core/base.py]: yeah see if u can catch something more specific bere

File: pandas/tests/groupby/test_groupby.py [hunk 5]
```diff
@@ -624,8 +624,13 @@ def test_as_index_series_return_frame(df):
     assert isinstance(result2, DataFrame)
     assert_frame_equal(result2, expected2)
 
-    # corner case
-    pytest.raises(Exception, grouped['C'].__getitem__, 'D')
+
+def test_as_index_series_column_slice_raises(df):
+    grouped = df.groupby('A', as_index=False)
```
Comment (jreback) [inline pandas/tests/groupby/test_groupby.py]: can you add the issue number here

File: pandas/core/base.py [hunk 6]
```diff
@@ -684,8 +684,14 @@ def _gotitem(self, key, ndim, subset=None):
         # with the same groupby
         kwargs = dict([(attr, getattr(self, attr))
                        for attr in self._attributes])
+
```
Comment (jreback) [inline pandas/core/base.py]: minor, but can you put the comments before the try (and consolidate them into 1)

**Review by jreback — APPROVED** [review]
very minor comments. pls rebase & merge on green.

**Comment by jreback** [comment]
@WillAyd can you rebase

**Comment by jreback** [comment]
thanks @WillAyd

**Pull request merged**